src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Statements.cs

```diff
@@ -411,16 +411,23 @@ private void Statement(Block block)
             if (Context.DesignTimeMode || !isMarkup)
             {
                 // CODE owns whitespace, MARKUP owns it ONLY in DesignTimeMode.
-                if (lastWs != null)
+                if (lastWhitespace != null)
                 {
-                    Accept(lastWs);
+                    Accept(lastWhitespace);
                 }
             }
             else
             {
+                var nextSymbol = Lookahead(1);
+
                 // MARKUP owns whitespace EXCEPT in DesignTimeMode.
                 PutCurrentBack();
-                PutBack(lastWs);
+
+                // Don't putback the whitespace if it precedes a '<text>' tag.
+                if (!nextSymbol.Content.Equals(SyntaxConstants.TextTagName))
+                {
+                    PutBack(lastWhitespace);
+                }
```

Comment:
I think you're right in 1 that there isn't really a need for whitespace before the `<text>` tag to be owned by markup. It probably had something to do with IntelliSense triggering (most of these weird whitespace things did :)) but I can't remember the details now. Double-check that IntelliSense works well around `<text>` tags and I'm happy with the change.